feat(xo-server-web-hooks): web hooks on XO API calls (#3155)
vatesfr/xen-orchestra: docs/web-hooks.md

# Web hooks

⚠ This feature is experimental!

## Configuration

The plugin "web-hooks" needs to be installed and loaded for this feature to work.

You can trigger an HTTP POST request to a URL when a Xen Orchestra API method is called.

* Go to Settings > Plugins > Web hooks
* Add new hooks
* For each hook, configure:
  * Method: the XO API method that will trigger the HTTP request when called
  * Type:
    * pre: the request will be sent when the method is called
    * post: the request will be sent after the method action is completed

>   * URL: the full URL the requests will be sent to

Reviewer: ```suggestion
  * URL: the full URL which the requests will be sent to
```

Reviewer 2: @badrAZ I'm not sure your correction is correct.

@Fohdeesha can you take a quick look at this documentation please?

Reviewer 3: grammar looks good with badrAZ's suggested change eg: "URL: the full URL which the requests will be sent to"

Reviewer 2: Can you take a full pass to this file (and make direct changes if you want to)?

And we'll ignore English related comments from other reviewers.

Reviewer 3: done, grammar edits made directly to source branch

* Save the plugin configuration

> From now on, every time the methods you configured will be called by an XO client, a request will be sent to the corresponding URLs.

Reviewer: ```suggestion
From now, a request will be sent to the corresponding URLs when a configured method is called by a XO client.
```

## Request content

```
POST / HTTP/1.1
Content-Type: application/json
```

The request's body is a JSON string representing an object with the following properties:

- `type`: `"pre"` or `"post"`
- `callId`: unique ID for this call to help match a pre-call and a post-call
- `userId`: unique internal ID of the user who performed the call
- `userName`: login/e-mail address of the user who performed the call
- `method`: name of the method that was called (e.g. `"vm.start"`)
- `params`: call parameters (object)
- `result`: call result on success ("post" hooks only)

> - `error`: call result on error ("post" hooks only)

Reviewer 2: What do think of adding a timestamp for post and pre, and the duration for post?
